File: perplexity-scraper/browser/nodejs/perplexity.mjs
// Perplexity scraper using the official @scrapeless-ai/sdk + puppeteer-core over CDP.
//
// Perplexity gates its answer pages behind Cloudflare and renders the
// input as a Lexical contenteditable, not a `<textarea>`. The scraper:
//   1. Warms up on the homepage to clear the Cloudflare cookie challenge.
//   2. Focuses `[role='textbox']`, types the prompt, presses Enter.
//   3. Waits for the URL to settle on `/search/<uuid>` and the answer
//      prose to render.
//   4. Extracts the question (`<h1 class="group/query …">`), the answer
//      (`<div class="prose …">`), and outbound citation `<a href>` links.

import { Scrapeless } from "@scrapeless-ai/sdk";
import puppeteer from "puppeteer-core";
import * as cheerio from "cheerio";

const DEFAULT_PROXY_COUNTRY = "US";
const DEFAULT_SESSION_TTL = 360;

const TRANSIENT_NET_ERRORS = [
  "ERR_TUNNEL_CONNECTION_FAILED",
  "ERR_CONNECTION_CLOSED",
  "ERR_CONNECTION_RESET",
  "ERR_CONNECTION_REFUSED",
  "ERR_TIMED_OUT",
  "ERR_NETWORK_CHANGED",
  "ERR_EMPTY_RESPONSE",
  "ERR_PROXY_CONNECTION_FAILED",
  "Navigation timeout",
  "net::",
];

function isTransientError(err) {
  const msg = String(err?.message ?? err ?? "");
  return TRANSIENT_NET_ERRORS.some((s) => msg.includes(s));
}

function requireKey() {
  const k = process.env.SCRAPELESS_API_KEY ?? process.env.SCRAPELESS_KEY;
  if (!k) throw new Error("SCRAPELESS_API_KEY env var not set. Sign up at https://app.scrapeless.com");
  return k;
}

function client() {
  return new Scrapeless({ apiKey: requireKey() });
}

async function withSessionRetry(fn, { retries = 2, label = "perplexity" } = {}) {
  let lastError;
  for (let attempt = 0; attempt <= retries; attempt++) {
    const { browserWSEndpoint } = await client().browser.create({
      proxyCountry: DEFAULT_PROXY_COUNTRY,
      sessionTTL: DEFAULT_SESSION_TTL,
    });
    let browser;
    try {
      browser = await puppeteer.connect({ browserWSEndpoint });
      const page = await browser.newPage();
      await page.setViewport({ width: 1366, height: 900 });
      return await fn(page, browser);
    } catch (e) {
      lastError = e;
      if (!isTransientError(e) && attempt > 0) throw e;
    } finally {
      try { await browser?.close(); } catch (_) {}
    }
    if (attempt < retries) await new Promise((r) => setTimeout(r, 4000 + attempt * 3000));
  }
  throw new Error(`${label}: giving up after ${retries + 1} attempts: ${lastError?.message ?? lastError}`);
}

async function warmupPerplexity(page) {
  await page.goto("https://www.perplexity.ai/", { waitUntil: "domcontentloaded", timeout: 60000 });
  await new Promise((r) => setTimeout(r, 8000));
}

async function submitPrompt(page, prompt) {
  await page.waitForSelector("[role='textbox'], [contenteditable='true']", { timeout: 25000 });
  await page.click("[role='textbox'], [contenteditable='true']");
  await page.keyboard.type(prompt, { delay: 30 });
  await new Promise((r) => setTimeout(r, 800));
  await page.keyboard.press("Enter");
}

async function waitForAnswer(page, { timeoutMs = 60000 } = {}) {
  // 1. URL settles on `/search/<uuid>`.
  const urlDeadline = Date.now() + 20000;
  while (Date.now() < urlDeadline) {
    if (/\/search\/[0-9a-f-]{8,}/i.test(page.url())) break;
    await new Promise((r) => setTimeout(r, 500));
  }
  // 2. Answer prose renders. Bail once it stops growing for ~3s or after the cap.
  const start = Date.now();
  let lastLen = 0;
  let lastChange = Date.now();
  while (Date.now() - start < timeoutMs) {
    await new Promise((r) => setTimeout(r, 800));
    let len = 0;
    try {
      len = await page.evaluate(() => {
        const el = document.querySelector("div.prose");
        return el ? (el.innerText || "").length : 0;
      });
    } catch (_) {}
    if (len !== lastLen) {
      lastLen = len;
      lastChange = Date.now();
    } else if (lastLen > 50 && Date.now() - lastChange >= 3000) {
      return;
    }
  }
}

// ---------------- scrape functions ----------------

export async function scrapeSearch(prompt) {
  return withSessionRetry(async (page) => {
    await warmupPerplexity(page);
    await submitPrompt(page, prompt);
    await waitForAnswer(page);
    const html = await page.content();
    const finalUrl = page.url();
    return parseSearch(html, prompt, finalUrl);
  }, { label: "scrape_search" });
}

// ---------------- parser ----------------

export function parseSearch(html, prompt, finalUrl) {
  const $ = cheerio.load(html);

  // Question — rendered inside the very first `h1.group/query` heading.
  // Cheerio's CSS uses dot-class so `.group\\/query` would otherwise need
  // an escaped slash; fall back to an attribute-contains match.
  let queryText = $("h1[class*='group/query']").first().text().trim();
  if (!queryText) queryText = prompt;
  queryText = queryText.replace(/\s+/g, " ").trim();

  // Answer — the first `<div class="prose ...">` block contains the
  // streamed response. Newer Perplexity layouts may render the answer in a
  // sibling `<div class="prose dark:prose-invert ...">` — match on `prose`.
  const proseNode = $("div[class*='prose']").first();
  const answerText = proseNode.text().replace(/\s+/g, " ").trim();

  // Citations — every external `<a href="https?://...">` that isn't a
  // perplexity-internal link or one of the layout chrome links. Deduped on
  // href.
  const seen = new Set();
  const citations = [];
  $("a[href^='http']").each((_, el) => {
    const href = $(el).attr("href") || "";
    if (!href || seen.has(href)) return;
    if (/perplexity\.ai/.test(href)) return;
    if (/\.perplexity\.ai\//.test(href)) return;
    if (/cloudflare\.com|gstatic\.com|twitter\.com\/PerplexityAI/.test(href)) return;
    seen.add(href);
    const title = $(el).text().replace(/\s+/g, " ").trim();
    citations.push({
      url: href,
      domain: domainOf(href),
      title,
    });
  });

  return {
    query: queryText,
    url: finalUrl,
    answer_text: answerText,
    citations,
  };
}

function domainOf(url) {
  try {
    let host = new URL(url).hostname || "";
    if (host.startsWith("www.")) host = host.slice(4);
    return host;
  } catch {
    return "";
  }
}

File: perplexity-scraper/browser/python/perplexity.py
"""Perplexity scraper using the official Scrapeless Python SDK + Playwright over CDP.

Perplexity gates its answer pages behind Cloudflare and renders the
prompt input as a Lexical contenteditable, not a ``<textarea>``. The
scraper:

1. Warms up on the homepage to clear the Cloudflare cookie challenge.
2. Focuses ``[role='textbox']``, types the prompt, presses Enter.
3. Waits for the URL to settle on ``/search/<uuid>`` and the answer
   prose to render.
4. Extracts the question (``<h1 class="group/query …">``), the answer
   (``<div class="prose …">``), and outbound citation ``<a href>`` links.
"""

from __future__ import annotations

import asyncio
import os
import re
from dataclasses import asdict, dataclass, field
from typing import Any
from urllib.parse import urlparse

from loguru import logger
from parsel import Selector
from playwright.async_api import async_playwright
from scrapeless import Scrapeless
from scrapeless.types import ICreateBrowser

DEFAULT_PROXY_COUNTRY = "US"
DEFAULT_SESSION_TTL = 360

TRANSIENT_NET_ERRORS = (
    "ERR_TUNNEL_CONNECTION_FAILED",
    "ERR_CONNECTION_CLOSED",
    "ERR_CONNECTION_RESET",
    "ERR_CONNECTION_REFUSED",
    "ERR_TIMED_OUT",
    "ERR_NETWORK_CHANGED",
    "ERR_EMPTY_RESPONSE",
    "ERR_PROXY_CONNECTION_FAILED",
    "Navigation timeout",
    "net::",
)


@dataclass
class Citation:
    url: str
    domain: str
    title: str


@dataclass
class Search:
    query: str
    url: str
    answer_text: str
    citations: list[Citation] = field(default_factory=list)


def _is_transient(err: Exception) -> bool:
    msg = str(err)
    return any(s in msg for s in TRANSIENT_NET_ERRORS)


def _client() -> Scrapeless:
    if not os.environ.get("SCRAPELESS_API_KEY") and not os.environ.get("SCRAPELESS_KEY"):
        raise RuntimeError(
            "SCRAPELESS_API_KEY env var not set. Sign up at https://app.scrapeless.com"
        )
    if not os.environ.get("SCRAPELESS_API_KEY") and os.environ.get("SCRAPELESS_KEY"):
        os.environ["SCRAPELESS_API_KEY"] = os.environ["SCRAPELESS_KEY"]
    return Scrapeless()


async def _with_session_retry(fn, *, retries: int = 2, label: str = "perplexity"):
    last_error: Exception | None = None
    for attempt in range(retries + 1):
        client = _client()
        session = client.browser.create(
            ICreateBrowser(proxy_country=DEFAULT_PROXY_COUNTRY, session_ttl=DEFAULT_SESSION_TTL)
        )
        async with async_playwright() as pw:
            browser = await pw.chromium.connect_over_cdp(session.browser_ws_endpoint)
            try:
                page = await browser.new_page(viewport={"width": 1366, "height": 900})
                return await fn(page)
            except Exception as e:  # noqa: BLE001
                last_error = e
                logger.warning("{} attempt {} failed: {}", label, attempt + 1, e)
                if not _is_transient(e) and attempt > 0:
                    raise
            finally:
                try:
                    await browser.close()
                except Exception:
                    pass
        if attempt < retries:
            await asyncio.sleep(4 + attempt * 3)
    raise RuntimeError(f"{label}: giving up after {retries + 1} attempts: {last_error}")


async def _warmup(page) -> None:
    await page.goto("https://www.perplexity.ai/", wait_until="domcontentloaded", timeout=60000)
    await asyncio.sleep(8)


async def _submit_prompt(page, prompt: str) -> None:
    await page.wait_for_selector("[role='textbox'], [contenteditable='true']", timeout=25000)
    await page.click("[role='textbox'], [contenteditable='true']")
    await page.keyboard.type(prompt, delay=30)
    await asyncio.sleep(0.8)
    await page.keyboard.press("Enter")


_URL_PAT = re.compile(r"/search/[0-9a-f-]{8,}", re.IGNORECASE)


async def _wait_for_answer(page, *, timeout_ms: int = 60000) -> None:
    """1. URL settles on `/search/<uuid>`. 2. Answer prose stops growing."""
    loop = asyncio.get_event_loop()
    url_deadline = loop.time() + 20
    while loop.time() < url_deadline:
        if _URL_PAT.search(page.url or ""):
            break
        await asyncio.sleep(0.5)

    start = loop.time()
    last_len = 0
    last_change = loop.time()
    while (loop.time() - start) * 1000 < timeout_ms:
        await asyncio.sleep(0.8)
        try:
            length = await page.evaluate(
                "() => { const el = document.querySelector('div.prose'); return el ? (el.innerText || '').length : 0; }"
            )
        except Exception:
            length = 0
        if length != last_len:
            last_len = length
            last_change = loop.time()
        elif last_len > 50 and (loop.time() - last_change) >= 3:
            return


# ---------------------------------------------------------------------------
# Scrape function
# ---------------------------------------------------------------------------


async def scrape_search(prompt: str) -> Search:
    async def _run(page):
        await _warmup(page)
        await _submit_prompt(page, prompt)
        await _wait_for_answer(page)
        html = await page.content()
        return parse_search(html, prompt, page.url)
    return await _with_session_retry(_run, label="scrape_search")


# ---------------------------------------------------------------------------
# Parser
# ---------------------------------------------------------------------------


_WS_RE = re.compile(r"\s+")
_PPLX_RE = re.compile(r"perplexity\.ai")
_BAD_HOST_RE = re.compile(r"cloudflare\.com|gstatic\.com|twitter\.com/PerplexityAI", re.IGNORECASE)


def _normalise(text: str) -> str:
    return _WS_RE.sub(" ", text or "").strip()


def _domain_of(url: str) -> str:
    try:
        host = (urlparse(url).hostname or "").lower()
        if host.startswith("www."):
            host = host[4:]
        return host
    except Exception:
        return ""


def parse_search(html: str, prompt: str, final_url: str) -> Search:
    sel = Selector(text=html)

    query_text = _normalise("".join(sel.css("h1[class*='group/query'] *::text").getall()))
    if not query_text:
        query_text = prompt

    answer_text = _normalise(" ".join(sel.css("div[class*='prose']").css("*::text").getall()))
    # First prose block only.
    proses = sel.css("div[class*='prose']")
    if proses:
        answer_text = _normalise(" ".join(proses[0].css("*::text").getall()))

    citations: list[Citation] = []
    seen: set[str] = set()
    for a in sel.css("a[href^='http']"):
        href = a.attrib.get("href", "")
        if not href or href in seen:
            continue
        if _PPLX_RE.search(href):
            continue
        if _BAD_HOST_RE.search(href):
            continue
        seen.add(href)
        title = _normalise("".join(a.css("*::text").getall()))
        citations.append(Citation(url=href, domain=_domain_of(href), title=title))

    return Search(
        query=query_text,
        url=final_url,
        answer_text=answer_text,
        citations=citations,
    )


def to_dict(obj: Any) -> Any:
    if hasattr(obj, "__dataclass_fields__"):
        return asdict(obj)
    if isinstance(obj, list):
        return [to_dict(x) for x in obj]
    return obj
